x-pack/elastic-agent/pkg/config/loader_test.go

```diff
@@ -0,0 +1,212 @@
+				filepath.Join("testdata", "inputs", "invalid-inputs.yml"),
+			},
+			inputsFolder: filepath.Join("testdata", "inputs", "*.yml"),
+			err:          true,
+		},
+		"two standalone configs can be merged without inputs": {
+			configs: []string{
+				filepath.Join("testdata", "standalone1.yml"),
+				filepath.Join("testdata", "standalone2.yml"),
+			},
+			inputsFolder: "",
+			expectedConfig: map[string]interface{}{
+				"outputs": map[string]interface{}{
+					"default": map[string]interface{}{
+						"type":    "elasticsearch",
+						"hosts":   []interface{}{"127.0.0.1:9201"},
+						"api-key": "my-secret-key",
+					},
+				},
+				"agent": map[string]interface{}{
+					"logging": map[string]interface{}{
+						"level": "debug",
+						"metrics": map[string]interface{}{
+							"enabled": false,
+						},
+					},
+				},
+			},
+		},
+		"one external config, standalone config without inputs section": {
+			configs: []string{
+				filepath.Join("testdata", "standalone1.yml"),
+				filepath.Join("testdata", "inputs", "log-inputs.yml"),
+				filepath.Join("testdata", "inputs", "metrics-inputs.yml"),
+			},
+			inputsFolder: filepath.Join("testdata", "inputs", "*.yml"),
+			expectedConfig: map[string]interface{}{
+				"outputs": map[string]interface{}{
+					"default": map[string]interface{}{
+						"type":    "elasticsearch",
+						"hosts":   []interface{}{"127.0.0.1:9201"},
+						"api-key": "my-secret-key",
+					},
+				},
+				"inputs": []interface{}{
+					map[string]interface{}{
+						"data_stream": map[string]interface{}{
+							"dataset": "system.auth",
+							"type":    "logs",
+						},
+						"exclude_files": []interface{}{".gz$"},
+						"id":            "logfile-system.auth-my-id",
+						"paths":         []interface{}{"/var/log/auth.log*", "/var/log/secure*"},
+						"use_output":    "default",
+					},
+					map[string]interface{}{
+						"data_stream": map[string]interface{}{
+							"dataset": "system.syslog",
+							"type":    "logs",
+						},
+						"type":          "logfile",
+						"id":            "logfile-system.syslog-my-id",
+						"exclude_files": []interface{}{".gz$"},
+						"paths":         []interface{}{"/var/log/messages*", "/var/log/syslog*"},
+						"use_output":    "default",
+					},
+					map[string]interface{}{
+						"data_stream": map[string]interface{}{
+							"dataset": "system.diskio",
+							"type":    "metrics",
+						},
+						"id":         "system/metrics-system.diskio-my-id",
+						"metricsets": []interface{}{"diskio"},
+						"period":     "10s",
+					},
+					map[string]interface{}{
+						"data_stream": map[string]interface{}{
+							"dataset": "system.filesystem",
+							"type":    "metrics",
+						},
+						"id":         "system/metrics-system.filesystem-my-id",
+						"metricsets": []interface{}{"filesystem"},
+						"period":     "30s",
+					},
+				},
+			},
+		},
+		"inputs sections of all external and standalone configuration are merged to the result": {
```

Comment:
What do you mean by conflict? Setting the same ID?

Reply:
> Do we need a test case where the external and standalone configurations both specify a logs-system.syslog-* datastream for example?

In this step of configuration loading, it is irrelevant, we do not need to test it. Parsing the `inputs` configuration is done separately in `pkg/composable`. In `pkg/config` I only do an initial sanity check to make sure that there are inputs in external configuration files.

> what happens if the following is duplicated between two configurations and the other fields do not match? For example if the use_output field were different.

The same thing happens if there is a conflict in the external and standalone configuration and there are multiple inputs with the same id in the standalone config because we are merging the configs. `use_output` is an odd one out. Agent starts separate Beats for every output, because each Beat can only send to one output. Hence, there is no problem because the ID in a given Beat instance is still unique.

However, if any other field is different, it means that there are inputs in the configuration of the same Beat instance that have the same ID. In Beats ID is only used to determine which state belongs to which input if the user is collecting from the same file/journal from multiple inputs. Setting the same ID for multiple inputs leads to undefined behaviour in those that require position tracking (e.g. `journald`, `log` or `filestream`) and the same sources are configured (e.g. reading from the same file or journal).

This configurations can work without issues:
```yaml
- type: filestream
  id: my-id
  paths: [/path/to/file]
- type: filestream
  id: my-id
  paths: [/path/to/other/file]
```

Problematic configuration:
```yaml
- type: filestream
  id: my-id
  paths: [/path/to/file]
- type: filestream
  id: my-id
  paths: [/path/to/file]
```

Unfortunately, at the moment we do not check if the IDs are unique in Beats. We just tell users to set unique IDs in the documentation. We could probably add a check to be more user friendly. I am find with addressing it, but in a separate follow-up PR.

Reply:
Thanks for the great explanation. 

If there are configurations that can lead to incorrect behaviour, then we should prevent them. I agree that can be a follow up PR, or even an issue in the backlog depending on how much work it would be to implement.

I'll mark this as resolved.